File: pet_vaccine_tracker.go
// pet_vaccine_tracker.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strconv"
	"time"
)

type Vaccine struct {
	Name        string    `json:"name"`
	Date        time.Time `json:"date"`
	ValidMonths int       `json:"valid_months"`
}

func (v Vaccine) ExpiryDate() time.Time {
	return v.Date.AddDate(0, v.ValidMonths, 0)
}

func (v Vaccine) DaysUntilExpiry() int {
	return int(v.ExpiryDate().Sub(time.Now()).Hours() / 24)
}

func (v Vaccine) IsExpired() bool {
	return v.DaysUntilExpiry() < 0
}

type Pet struct {
	ID        int       `json:"id"`
	Name      string    `json:"name"`
	Species   string    `json:"species"`
	Birthdate time.Time `json:"birthdate"`
	Owner     string    `json:"owner"`
	Vaccines  []Vaccine `json:"vaccines"`
}

type Tracker struct {
	Pets []Pet `json:"pets"`
	File string
}

func NewTracker(file string) *Tracker {
	t := &Tracker{File: file}
	t.load()
	return t
}

func (t *Tracker) load() {
	data, err := os.ReadFile(t.File)
	if err != nil {
		return
	}
	json.Unmarshal(data, t)
}

func (t *Tracker) save() {
	data, _ := json.MarshalIndent(t, "", "  ")
	os.WriteFile(t.File, data, 0644)
}

func (t *Tracker) nextID() int {
	max := 0
	for _, p := range t.Pets {
		if p.ID > max {
			max = p.ID
		}
	}
	return max + 1
}

func (t *Tracker) getPet(id int) *Pet {
	for i, p := range t.Pets {
		if p.ID == id {
			return &t.Pets[i]
		}
	}
	return nil
}

func (t *Tracker) AddPet(name, species, birthdate, owner string) {
	birth, _ := time.Parse("2006-01-02", birthdate)
	pet := Pet{
		ID:        t.nextID(),
		Name:      name,
		Species:   species,
		Birthdate: birth,
		Owner:     owner,
	}
	t.Pets = append(t.Pets, pet)
	t.save()
	fmt.Printf("✅ Pet added: %s (ID: %d)\n", name, pet.ID)
}

func (t *Tracker) AddVaccine(petID int, vaccineName, date string, validMonths int) {
	pet := t.getPet(petID)
	if pet == nil {
		fmt.Printf("❌ Pet with ID %d not found.\n", petID)
		return
	}
	vDate, _ := time.Parse("2006-01-02", date)
	v := Vaccine{Name: vaccineName, Date: vDate, ValidMonths: validMonths}
	pet.Vaccines = append(pet.Vaccines, v)
	t.save()
	fmt.Printf("💉 Vaccine '%s' added for %s.\n", vaccineName, pet.Name)
}

func (t *Tracker) ListPets() {
	if len(t.Pets) == 0 {
		fmt.Println("No pets.")
		return
	}
	fmt.Println("\n🐾 Pets:")
	for _, p := range t.Pets {
		fmt.Printf("  [%d] %s (%s) – Owner: %s\n", p.ID, p.Name, p.Species, p.Owner)
	}
}

func (t *Tracker) ShowPet(petID int) {
	pet := t.getPet(petID)
	if pet == nil {
		fmt.Printf("❌ Pet with ID %d not found.\n", petID)
		return
	}
	fmt.Printf("\n🐶 %s (%s)\n", pet.Name, pet.Species)
	fmt.Printf("  Owner: %s\n", pet.Owner)
	fmt.Printf("  Birth: %s\n", pet.Birthdate.Format("2006-01-02"))
	if len(pet.Vaccines) == 0 {
		fmt.Println("  No vaccines recorded.")
		return
	}
	fmt.Println("  Vaccines:")
	for i, v := range pet.Vaccines {
		days := v.DaysUntilExpiry()
		var status string
		if days < 0 {
			status = fmt.Sprintf("\033[31mOVERDUE by %d days 🚨\033[0m", -days)
		} else if days <= 30 {
			status = fmt.Sprintf("\033[33mexpires in %d days ⚠️\033[0m", days)
		} else {
			status = fmt.Sprintf("\033[32mexpires in %d days\033[0m", days)
		}
		fmt.Printf("    [%d] %s – %s – %s\n", i+1, v.Name, v.Date.Format("2006-01-02"), status)
	}
}

func (t *Tracker) CheckExpiring(daysThreshold int) {
	var expiring []struct {
		pet  *Pet
		v    Vaccine
		days int
	}
	for i := range t.Pets {
		for _, v := range t.Pets[i].Vaccines {
			days := v.DaysUntilExpiry()
			if days >= 0 && days <= daysThreshold {
				expiring = append(expiring, struct {
					pet  *Pet
					v    Vaccine
					days int
				}{&t.Pets[i], v, days})
			} else if days < 0 {
				expiring = append(expiring, struct {
					pet  *Pet
					v    Vaccine
					days int
				}{&t.Pets[i], v, days})
			}
		}
	}
	if len(expiring) == 0 {
		fmt.Println("✅ No vaccines expiring soon.")
		return
	}
	fmt.Printf("\n⏰ Vaccines expiring within %d days (or overdue):\n", daysThreshold)
	for _, e := range expiring {
		var status string
		if e.days < 0 {
			status = fmt.Sprintf("\033[31mOVERDUE by %d days\033[0m", -e.days)
		} else {
			status = fmt.Sprintf("\033[33m%d days left\033[0m", e.days)
		}
		fmt.Printf("  %s – %s – %s\n", e.pet.Name, e.v.Name, status)
	}
}

func main() {
	addPetCmd := flag.NewFlagSet("add-pet", flag.ExitOnError)
	addPetName := addPetCmd.String("name", "", "")
	addPetSpecies := addPetCmd.String("species", "", "")
	addPetBirth := addPetCmd.String("birthdate", "", "")
	addPetOwner := addPetCmd.String("owner", "", "")

	addVaccCmd := flag.NewFlagSet("add-vaccine", flag.ExitOnError)
	addVaccPetID := addVaccCmd.Int("pet-id", 0, "")
	addVaccName := addVaccCmd.String("vaccine-name", "", "")
	addVaccDate := addVaccCmd.String("date", "", "")
	addVaccMonths := addVaccCmd.Int("valid-months", 0, "")

	listCmd := flag.NewFlagSet("list", flag.ExitOnError)
	showCmd := flag.NewFlagSet("show", flag.ExitOnError)
	showPetID := showCmd.Int("pet-id", 0, "")
	checkCmd := flag.NewFlagSet("check", flag.ExitOnError)
	checkDays := checkCmd.Int("days", 30, "")

	if len(os.Args) < 2 {
		fmt.Println("Usage: pet_vaccine_tracker <command> [options]")
		return
	}
	tracker := NewTracker("pets.json")

	switch os.Args[1] {
	case "add-pet":
		addPetCmd.Parse(os.Args[2:])
		tracker.AddPet(*addPetName, *addPetSpecies, *addPetBirth, *addPetOwner)
	case "add-vaccine":
		addVaccCmd.Parse(os.Args[2:])
		tracker.AddVaccine(*addVaccPetID, *addVaccName, *addVaccDate, *addVaccMonths)
	case "list":
		listCmd.Parse(os.Args[2:])
		tracker.ListPets()
	case "show":
		showCmd.Parse(os.Args[2:])
		tracker.ShowPet(*showPetID)
	case "check":
		checkCmd.Parse(os.Args[2:])
		tracker.CheckExpiring(*checkDays)
	default:
		fmt.Println("Unknown command")
	}
}

File: pet_vaccine_tracker.py
# pet_vaccine_tracker.py
import sys, os, json, argparse
from datetime import datetime, timedelta
from typing import List, Dict, Optional

try:
    from colorama import init, Fore, Style
    init()
    COLORS = True
except ImportError:
    COLORS = False
    Fore = Style = type('', (), {'RESET_ALL':'', 'GREEN':'', 'RED':'', 'YELLOW':'', 'CYAN':''})()

DATA_FILE = "pets.json"

class Vaccine:
    def __init__(self, name: str, date: str, valid_months: int):
        self.name = name
        self.date = datetime.strptime(date, "%Y-%m-%d")
        self.valid_months = valid_months

    def expiry_date(self):
        return self.date + timedelta(days=30*self.valid_months)

    def days_until_expiry(self):
        return (self.expiry_date() - datetime.now()).days

    def is_expired(self):
        return self.days_until_expiry() < 0

    def to_dict(self):
        return {"name": self.name, "date": self.date.isoformat(), "valid_months": self.valid_months}

    @classmethod
    def from_dict(cls, d):
        return cls(d["name"], d["date"], d["valid_months"])

class Pet:
    def __init__(self, name: str, species: str, birthdate: str, owner: str, pet_id: Optional[int] = None):
        self.id = pet_id
        self.name = name
        self.species = species
        self.birthdate = datetime.strptime(birthdate, "%Y-%m-%d")
        self.owner = owner
        self.vaccines: List[Vaccine] = []

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "species": self.species,
            "birthdate": self.birthdate.isoformat(),
            "owner": self.owner,
            "vaccines": [v.to_dict() for v in self.vaccines]
        }

    @classmethod
    def from_dict(cls, d):
        p = cls(d["name"], d["species"], d["birthdate"], d["owner"], d["id"])
        p.vaccines = [Vaccine.from_dict(v) for v in d.get("vaccines", [])]
        return p

class Tracker:
    def __init__(self):
        self.pets: List[Pet] = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.pets = [Pet.from_dict(p) for p in data]

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump([p.to_dict() for p in self.pets], f, indent=2)

    def next_id(self):
        return max([p.id for p in self.pets] + [0]) + 1

    def get_pet(self, pet_id: int) -> Optional[Pet]:
        for p in self.pets:
            if p.id == pet_id:
                return p
        return None

    def add_pet(self, name, species, birthdate, owner):
        pet = Pet(name, species, birthdate, owner, self.next_id())
        self.pets.append(pet)
        self.save()
        print(f"✅ Pet added: {name} (ID: {pet.id})")

    def add_vaccine(self, pet_id, vaccine_name, date, valid_months):
        pet = self.get_pet(pet_id)
        if not pet:
            print(f"❌ Pet with ID {pet_id} not found.")
            return
        v = Vaccine(vaccine_name, date, valid_months)
        pet.vaccines.append(v)
        self.save()
        print(f"💉 Vaccine '{vaccine_name}' added for {pet.name}.")

    def list_pets(self):
        if not self.pets:
            print("No pets.")
            return
        print(f"\n🐾 {Fore.CYAN}Pets{Style.RESET_ALL}")
        for p in self.pets:
            print(f"  {Fore.YELLOW}[{p.id}]{Style.RESET_ALL} {p.name} ({p.species}) – Owner: {p.owner}")

    def show_pet(self, pet_id):
        pet = self.get_pet(pet_id)
        if not pet:
            print(f"❌ Pet with ID {pet_id} not found.")
            return
        print(f"\n🐶 {Fore.CYAN}{pet.name}{Style.RESET_ALL} ({pet.species})")
        print(f"  Owner: {pet.owner}")
        print(f"  Birth: {pet.birthdate.strftime('%Y-%m-%d')}")
        if not pet.vaccines:
            print("  No vaccines recorded.")
            return
        print("  Vaccines:")
        for i, v in enumerate(pet.vaccines, 1):
            days = v.days_until_expiry()
            if days < 0:
                status = f"{Fore.RED}OVERDUE by {-days} days 🚨{Style.RESET_ALL}"
            elif days <= 30:
                status = f"{Fore.YELLOW}expires in {days} days ⚠️{Style.RESET_ALL}"
            else:
                status = f"{Fore.GREEN}expires in {days} days{Style.RESET_ALL}"
            print(f"    [{i}] {v.name} – {v.date.strftime('%Y-%m-%d')} – {status}")

    def check_expiring(self, days_threshold=30):
        expiring = []
        for p in self.pets:
            for v in p.vaccines:
                days = v.days_until_expiry()
                if 0 <= days <= days_threshold:
                    expiring.append((p, v, days))
                elif days < 0:
                    expiring.append((p, v, days))  # overdue
        if not expiring:
            print("✅ No vaccines expiring soon.")
            return
        print(f"\n⏰ {Fore.CYAN}Vaccines expiring within {days_threshold} days (or overdue){Style.RESET_ALL}")
        for p, v, days in expiring:
            if days < 0:
                status = f"{Fore.RED}OVERDUE by {-days} days{Style.RESET_ALL}"
            else:
                status = f"{Fore.YELLOW}{days} days left{Style.RESET_ALL}"
            print(f"  {p.name} – {v.name} – {status}")

def main():
    parser = argparse.ArgumentParser(description="Pet Vaccination Tracker")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    add_pet = subparsers.add_parser("add-pet")
    add_pet.add_argument("name")
    add_pet.add_argument("species")
    add_pet.add_argument("birthdate")
    add_pet.add_argument("owner")

    add_vacc = subparsers.add_parser("add-vaccine")
    add_vacc.add_argument("pet_id", type=int)
    add_vacc.add_argument("vaccine_name")
    add_vacc.add_argument("date")
    add_vacc.add_argument("valid_months", type=int)

    subparsers.add_parser("list")
    show_parser = subparsers.add_parser("show")
    show_parser.add_argument("pet_id", type=int)

    check_parser = subparsers.add_parser("check")
    check_parser.add_argument("--days", type=int, default=30)

    args = parser.parse_args()
    tracker = Tracker()

    if args.cmd == "add-pet":
        tracker.add_pet(args.name, args.species, args.birthdate, args.owner)
    elif args.cmd == "add-vaccine":
        tracker.add_vaccine(args.pet_id, args.vaccine_name, args.date, args.valid_months)
    elif args.cmd == "list":
        tracker.list_pets()
    elif args.cmd == "show":
        tracker.show_pet(args.pet_id)
    elif args.cmd == "check":
        tracker.check_expiring(args.days)

if __name__ == "__main__":
    main()
